components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/PostAuthnMissingChallengeQuestionsHandler.java

```diff
@@ -0,0 +1,411 @@
+        }
+        return false;
+    }
+
+    /**
+     * Returns a list of challenge questions for a given user.
+     *
+     * @param user Authenticated User.
+     * @return List of ChallengeQuestions.
+     */
+    private List<ChallengeQuestion> getChallengeQuestions(AuthenticatedUser user) {
+        String tenantDomain = MultitenantUtils.getTenantDomain(user.getUserName());
+
+        try {
+            return ChallengeQuestionManager.getInstance().getAllChallengeQuestions
+                    (tenantDomain);
+        } catch (IdentityRecoveryServerException e) {
+            log.error("Identity recovery server error occurred for user:" + user.getUserName(), e);
+            return null;
+        }
+    }
+
+    /**
+     * Set the challenge questions for the user by calling ChallengeQuestionManger.
+     *
+     * @param user                 Authenticated User.
+     * @param userChallengeAnswers Array of UserChallengeAnswer
+     */
+    private void setChallengeQuestionAnswers(User user, UserChallengeAnswer[] userChallengeAnswers) {
+
+        try {
+            ChallengeQuestionManager.getInstance().setChallengesOfUser(user, userChallengeAnswers);
+        } catch (IdentityRecoveryException e) {
+            log.error("Unable to save challenge question answers", e);
+        }
+    }
+
+    /**
+     * Returns array of UserChallengeAnswer from the servlet request parameters from the front-end
+     *
+     * @param servletRequest HTTP Servlet Request.
+     * @return Array of UserChallengeAnswer.
+     */
+    private UserChallengeAnswer[] retrieveChallengeQuestionAnswers(HttpServletRequest servletRequest,
+                                                                   List<ChallengeQuestion> list) {
+        Map<String, String> questionsMap = new HashMap<>();
+        Map<String, String> answersMap = new HashMap<>();
+        List<UserChallengeAnswer> questionsAndAnswers = new ArrayList<>();
+        List<String> paramNamesList = new ArrayList<>();
+
+        Enumeration<String> paramNames = servletRequest.getParameterNames();
+        while (paramNames.hasMoreElements()) {
+            paramNamesList.add(paramNames.nextElement());
+        }
+
+        for (String requestParam : paramNamesList) {
+            if (requestParam.contains(SELECTED_CHALLENGE_QUESTION_PREFIX)) {
+                String question = servletRequest.getParameter(requestParam);
+                String questionSetID = requestParam.replace(SELECTED_CHALLENGE_QUESTION_PREFIX, "");
+                questionsMap.put(questionSetID, question);
+
+            } else if (requestParam.contains(CHALLENGE_QUESTION_ANSWER_PREFIX)) {
+                String answer = servletRequest.getParameter(requestParam);
+                String answerSetID = requestParam.replace(CHALLENGE_QUESTION_ANSWER_PREFIX, "");
+                answersMap.put(answerSetID, answer);
+            }
+        }
+        for (String s : questionsMap.keySet()) {
```

Comment:
Better to have a meaningful variable name.

Reply:
fixed